zaza/model.py

```diff
@@ -1316,6 +1322,32 @@ async def _block():
 block_until = sync_wrapper(async_block_until)
 
 
+def file_contents(unit_name, path, timeout=None):
+    """Return the contents of a file.
+
+    :param path: File path
+    :param unit_name: Unit name, either appname/N or appname/leader
+    :type unit_name: str
+    :param timeout: Timeout in seconds
+    :type timeout: float
+    :returns: File contents
+    :rtype: str
+    """
+    app, unit_id = unit_name.split("/")
+    if unit_id == "leader":
+        target = app
+        run_func = run_on_leader
+    else:
+        target = unit_name
+        run_func = run_on_unit
+    cmd = "cat {}".format(path)
+    result = run_func(target, cmd, timeout=timeout)
+    err = result.get("Stderr")
+    if err:
+        raise RemoteFileError(err)
+    return result.get("Stdout")
```

Comment:
Isn't there a Code that indicates an error too?  The `RemoteFileError` might be more useful if it also contained the error code from the cat?

Reply:
I'm a bit confused as to why there are sync and async functions in this PR?  I'm a bit worried that `file_contents()` is being used in an async test function?  It just seems a bit strange.  Please could you show how these are being used?  Thanks.

Reply:
> Isn't there a Code that indicates an error too? The `RemoteFileError` might be more useful if it also contained the error code from the cat?

The `cat` exit codes aren't terribly interesting -- they're 0 on success and 1 on failure. Do you think it's worthwhile including them?

Reply:
> I'm a bit confused as to why there are sync and async functions in this PR? I'm a bit worried that `file_contents()` is being used in an async test function? It just seems a bit strange. Please could you show how these are being used? Thanks.

I've provided an async function for blocking on regexp as well as I thought there might be a usecase for expensive regexp matching. I didn't think there would be much use for an async `file_contents()`

Do you want me to provide functional testing for those?

Reply:
> > Isn't there a Code that indicates an error too? The `RemoteFileError` might be more useful if it also contained the error code from the cat?
> 
> The `cat` exit codes aren't terribly interesting -- they're 0 on success and 1 on failure. Do you think it's worthwhile including them?

Actually, no, in that case, it's not terribly useful.  You made the right call on that.

Reply:
> > I'm a bit confused as to why there are sync and async functions in this PR? I'm a bit worried that `file_contents()` is being used in an async test function? It just seems a bit strange. Please could you show how these are being used? Thanks.
> 
> I've provided an async function for blocking on regexp as well as I thought there might be a usecase for expensive regexp matching. I didn't think there would be much use for an async `file_contents()`
> 
> Do you want me to provide functional testing for those?

If I understand correctly, there isn't actually a use-case for `file_contents()`?  i.e. there isn't a test in zaza-openstack-tests that uses it?  In that case, there is no need to add it here.  This is affectionately termed 'gold-plating' ... the code has to be maintained, but nothing is using it, and it's been added 'just-in-case'.

Is this *also* the case for the other functions in this PR?

If I've misunderstood, my apologies!

Reply:
I meant to say that I didn't see a use case for an _async_ version of `file_contents()` -- only for the synchronous version. If so desired I could also create an async version of `file_contents()` though.

The functions added here indeed come from a concrete use case. I've used their predecessor versions in the launchpad MR [here](https://code.launchpad.net/~peter-sabaini/charm-juju-lint/+git/charm-juju-lint/+merge/394000). 

In short: 

- the `file_contents()` is used to get values from a test dependency which are then used to configure the system. The test dependency is a juju controller created within the test model, from where I need to retrieve the generated model parameters to then feed them into the Juju lint charm.

- `block_until_file_matches_re()` is used to test for SUT output which is better described as a pattern, as there are some varying bits in there which aren't relevant for the test. The test output being from the auto-lint program where I'd like to match on the output keywords but not on the generated hostname, time stamps etc.

I hope this clarifies? If not feel free to also ping me on IRC